app/vmctl/remoteread/remoteread.go

```diff
@@ -0,0 +1,263 @@
+		err := c.fetch(ctx, b, streamCb)
+		if err != nil {
+			if errors.Is(err, context.Canceled) {
+				return fmt.Errorf("process stoped")
+			}
+			logger.Errorf("attempt %d to fetch data from remote storage: %s", i+1, err)
+			// sleeping to avoid remote db hammering
+			time.Sleep(time.Second)
+			continue
+		}
+		return nil
+	}
+	return nil
+}
+
+// Ping checks the health of the read source
+func (c *Client) Ping() error {
+	url := c.addr + healthPath
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return fmt.Errorf("cannot create request to %q: %s", url, err)
+	}
+	if c.user != "" {
+		req.SetBasicAuth(c.user, c.password)
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("bad status code: %d", resp.StatusCode)
+	}
+	return nil
+}
+
+func (c *Client) fetch(ctx context.Context, data []byte, streamCb StreamCallback) error {
+	r := bytes.NewReader(data)
+	url := c.addr + remoteReadPath
+	req, err := http.NewRequest("POST", url, r)
+	if err != nil {
+		return fmt.Errorf("failed to create new HTTP request: %w", err)
+	}
+
+	req.Header.Add("Content-Encoding", "snappy")
+	req.Header.Add("Accept-Encoding", "snappy")
+	req.Header.Set("Content-Type", "application/x-protobuf")
+	req.Header.Set("X-Prometheus-Remote-Read-Version", "0.1.0")
+
+	if c.user != "" {
+		req.SetBasicAuth(c.user, c.password)
+	}
+
+	resp, err := c.c.Do(req.WithContext(ctx))
+	if err != nil {
+		return fmt.Errorf("error while sending request to %s: %w; Data len %d(%d)",
+			req.URL.Redacted(), err, len(data), r.Size())
+	}
+	defer func() { _ = resp.Body.Close() }()
+	if resp.StatusCode != http.StatusNoContent && resp.StatusCode != http.StatusOK {
+		body, _ := io.ReadAll(resp.Body)
+		return fmt.Errorf("unexpected response code %d for %s. Response body %q",
+			resp.StatusCode, req.URL.Redacted(), body)
+	}
+	defer func() { _ = resp.Body.Close() }()
+
+	bb := bodyBufferPool.Get()
+	defer bodyBufferPool.Put(bb)
+
+	stream := remote.NewChunkedReader(resp.Body, remote.DefaultChunkedReadLimit, bb.B)
+
+	for {
+		var ts prompb.TimeSeries
+		res := &prompb.ChunkedReadResponse{}
+		err := stream.NextProto(res)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		for _, series := range res.ChunkedSeries {
+			ts.Labels = append(ts.Labels, series.Labels...)
+			for _, chunk := range series.Chunks {
+				samples, err := parseSamples(chunk.Data)
+				if err != nil {
+					return err
+				}
+				ts.Samples = append(ts.Samples, samples...)
+			}
+			if err := streamCb(ts); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (c *Client) query(filter *Filter) (*prompb.Query, error) {
+	var ms *labels.Matcher
+	if filter.Label == "" && filter.LabelValue == "" {
+		ms = labels.MustNewMatcher(labels.MatchRegexp, labels.MetricName, ".+")
+	} else {
+		ms = labels.MustNewMatcher(labels.MatchRegexp, filter.Label, filter.LabelValue)
```

Comment:
NewMatcher, MustNewMatcher - may panic with user input.

Reply:
NewMatcher does not panic it returns an error.

Reply:
updated logic